File: py/merge_sort.py
def merge(arr, l, m, r):

    #0..3 -> 4, 1..4 ->4
    l1= m-l+1
    a1 = [0]*(l1)
    for i in range(l1):
        a1[i] = arr[l+i]

    #(3+1)..8 -> 5
    l2=r-m
    a2 = [0]*l2
    for i in range(l2):
        a2[i] = arr[m+1+i]

    i=0 #index for 1st half
    j=0 #index for 2nd half
    k=l #index for main array

    while i<l1 and j<l2:

        if a1[i] <= a2[j]:
            arr[k]= a1[i]
            i+=1
        else:
            arr[k]=a2[j]
            j+=1

        k+=1

    while i<l1:
        arr[k] = a1[i]
        k+=1
        i+=1

    while j<l2:
        arr[k] = a2[j]
        k+=1
        j+=1


def merge_sort(arr, l, r):

    if l<r:
        mid = l+(r-l)/2
        merge_sort(arr,l,mid)
        merge_sort(arr,mid+1,r)
        merge(arr,l,mid,r)

if __name__ == '__main__':

    arr =[2,6,3,4,1,2,9,5,8]
    #passing the index
    merge_sort(arr,0, len(arr)-1)

    print arr

File: go/sorting/merge-sort.go
package main

import "fmt"

func merge(arr []int, lidx int, mid int, ridx int) {

	n1 := mid - lidx + 1
	n2 := ridx - mid

	ar1 := make([]int, n1)
	ar2 := make([]int, n2)

	for idx, _ := range ar1 {
		ar1[idx] = arr[lidx+idx]
	}

	for idx, _ := range ar2 {
		ar2[idx] = arr[mid+1+idx]
	}

	//copy(ar1, arr[lidx:mid])
	//copy(ar2, arr[mid+1:ridx])

	l_ptr := 0
	r_ptr := 0
	ptr := lidx

	for l_ptr < n1 && r_ptr < n2 {

		if ar1[l_ptr] < ar2[r_ptr] {
			arr[ptr] = ar1[l_ptr]
			l_ptr++
		} else {
			arr[ptr] = ar2[r_ptr]
			r_ptr++
		}
		ptr++
	}

	if l_ptr >= n1 {
		for r_ptr < n2 {
			arr[ptr] = ar2[r_ptr]
			ptr++
			r_ptr++
		}
	}

	if r_ptr >= n2 {
		for l_ptr < n1 {
			arr[ptr] = ar1[l_ptr]
			ptr++
			l_ptr++
		}
	}
}

func mergesort(arr []int, lidx int, ridx int) {

	if lidx < ridx {
		mid := lidx + ((ridx - lidx) / 2)

		mergesort(arr[:], lidx, mid)
		mergesort(arr[:], mid+1, ridx)

		merge(arr[:], lidx, mid, ridx)
	}
}

func main() {

	arr := [11]int{1, 4, 16, 29, 46, 53, 67, 88, 56, 43, 39}

	mergesort(arr[:], 0, len(arr)-1)

	fmt.Println(arr)

}
